Add Support for changing Join and Leave Messages

MojangPlsFix opened the pull request:
**Current Status:**
Currently there is no way on the server to change the join or leave messages of a player (As mentioned [here](https://forums.minecraftforge.net/topic/92256-1163-make-custom-join-messages/))

**My Changes:**
This Pull Request adds 2 new Events `LeaveMessageFormat` and `JoinMessageFormat`, these 2 events behave like the `TabListNameFormat` Event. If the message in the event remains null, then the default behaviour, that was before (Yellow Text joined/left the game). If not null, then the Component is broadcasted to all players instead of the default message.

**Why?**
Our Minecraft Network wants to make our Forge servers only forge, so no extra server software. When I was porting our plugins to forge mods, I encountered the problem, that I cannot disable Join/Leave Messages or even change them.  
Adding this feature would benefit server owners for making server side mods and also adds somekind of customization without changing the client.

MojangPlsFix committed:
Add Support for changing join and leave Messages

MojangPlsFix committed:
fix javadoc for JoinMessageFormat and LeaveMessageFormat

MojangPlsFix committed:
use getPlayerLeaveMessage instead of getPlayerJoinMessage

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/MinecraftForge/MinecraftForge?pullRequest=8311) <br/>All committers have signed the CLA.

KnightMiner commented:
Your "why" for the pull request says "because I cannot change the messages". Why do you want to change the message? What is an example changed message? I feel like a server resource pack would be sufficient for this.

MojangPlsFix commented:
My Change also allows you to send no message at all. Which is what I want in my case. We have servers with 20+ players and they are constantly joining and leaving the server. This can really spam the chat. You can compare it with the Forge Discord Server, do you want to see everyone who joins or leaves the discord server in the general chat? 

One another use case of my change could be changing the join messages based on who joins the server. e.g. If a op joins, then the text is red or show his/her rank in the join message. 

Also using a resource pack is not a really clean way to do that, if you have a bigger server. The user has to download the pack, in my knowledge. Not everyone wants that or has it enabled.

ChampionAsh5357 commented on patches/minecraft/net/minecraft/server/network/ServerGamePacketListenerImpl.java.patch:
```diff
@@ -32,6 +32,16 @@
+    public void m_7026_(Component p_9825_) {
+       f_9744_.info("{} lost connection: {}", this.f_9743_.m_7755_().getString(), p_9825_.getString());
+       this.f_9745_.m_129929_();
++      net.minecraft.network.chat.MutableComponent mutableComponent = net.minecraftforge.event.ForgeEventFactory.getPlayerLeaveMessage( this.f_9743_ );
++      if(mutableComponent == null) mutableComponent = new TranslatableComponent("multiplayer.player.left", this.f_9743_.m_5446_()).m_130940_(ChatFormatting.YELLOW);
++      if(!mutableComponent.getString().isEmpty()) this.f_9745_.m_6846_().m_11264_(mutableComponent, ChatType.SYSTEM, Util.f_137441_);
+       this.f_9745_.m_6846_().m_11264_((new TranslatableComponent("multiplayer.player.left", this.f_9743_.m_5446_())).m_130940_(ChatFormatting.YELLOW), ChatType.SYSTEM, Util.f_137441_);
```
Sending the message twice. Also, I'm pretty sure this could be compressed down to a single line to a forge hook.

MojangPlsFix replied:
Is now a one Liner using ForgeHooks. Hope this is the correct way implementing this.

ChampionAsh5357 commented on patches/minecraft/net/minecraft/server/players/PlayerList.java.patch:
```diff
@@ -24,9 +24,30 @@
+       p_11263_.m_8952_().m_12789_(p_11263_);
+       this.m_11273_(serverlevel1.m_6188_(), p_11263_);
+       this.f_11195_.m_129929_();
+-      MutableComponent mutablecomponent;
+-      if (p_11263_.m_36316_().getName().equalsIgnoreCase(s)) {
+-         mutablecomponent = new TranslatableComponent("multiplayer.player.joined", p_11263_.m_5446_());
+-      } else {
+-         mutablecomponent = new TranslatableComponent("multiplayer.player.joined.renamed", p_11263_.m_5446_(), s);
+-      }
+-
+-      this.m_11264_(mutablecomponent.m_130940_(ChatFormatting.YELLOW), ChatType.SYSTEM, Util.f_137441_);
++      MutableComponent mutablecomponent = net.minecraftforge.event.ForgeEventFactory.getPlayerJoinMessage( p_11263_ );
++      if(mutablecomponent == null) {
++         if (p_11263_.m_36316_().getName().equalsIgnoreCase(s)) {
++            mutablecomponent = new TranslatableComponent("multiplayer.player.joined", p_11263_.m_5446_());
++         } else {
++            mutablecomponent = new TranslatableComponent("multiplayer.player.joined.renamed", p_11263_.m_5446_(), s);
++         }
++         mutablecomponent = mutablecomponent.m_130940_(ChatFormatting.YELLOW);
++      }
++      if(!mutablecomponent.getString().isEmpty()) {
++         this.m_11264_(mutablecomponent, ChatType.SYSTEM, Util.f_137441_);
++      }
```
Can be compressed down considerably using a forge hook.

MojangPlsFix replied:
It is now a one line like the LeaveFormat.

ChampionAsh5357 commented on src/test/resources/META-INF/mods.toml:
```diff
@@ -172,5 +172,6 @@ license="LGPL v2.1"
     modId="part_entity_test"
 [[mods]]
     modId="capabilities_test"
-
```
Don't need to remove the new line at the end of the file.

MojangPlsFix replied:
did not think of that. I added it back

ChampionAsh5357 reviewed (changes requested):
> My Change also allows you to send no message at all. Which is what I want in my case.

If you wanted to send no message, you should make the event cancelable. Then just don't send anything to the clients. Why send an empty mutable component if it won't be displayed anyways? It will just be an empty line submitted to all clients.

MojangPlsFix committed:
add new line back

MojangPlsFix committed:
make JoinMessageFormat and LeaveMessageFormat cancelable

MojangPlsFix committed:
handle canceled events

MojangPlsFix committed:
decrease size of patch for JoinMessageFormat using ForgeHooks

MojangPlsFix committed:
make compress Patch with ForgeHooks

MojangPlsFix commented:
Also, if this pr is accepted, can I also backport the feature to 1.16.5 or is this too much of a change, for the LTS version?

MojangPlsFix commented:
> If you wanted to send no message, you should make the event cancelable. Then just don't send anything to the clients. Why send an empty mutable component if it won't be displayed anyways? It will just be an empty line submitted to all clients.

The Events are now cancelable. Wanted to do the same behaviour as the NameFormat Event, but did not think of that first.

ChampionAsh5357 commented on patches/minecraft/net/minecraft/server/network/ServerGamePacketListenerImpl.java.patch:
```diff
@@ -32,6 +32,15 @@
+    public void m_7026_(Component p_9825_) {
+       f_9744_.info("{} lost connection: {}", this.f_9743_.m_7755_().getString(), p_9825_.getString());
+       this.f_9745_.m_129929_();
+-      this.f_9745_.m_6846_().m_11264_((new TranslatableComponent("multiplayer.player.left", this.f_9743_.m_5446_())).m_130940_(ChatFormatting.YELLOW), ChatType.SYSTEM, Util.f_137441_);
++      net.minecraftforge.common.ForgeHooks.onLeaveMessageFormat(this.f_9743_,this.f_9745_);
```
The forge hook is responsible for providing the message, not sending it as well.

ChampionAsh5357 commented on patches/minecraft/net/minecraft/server/players/PlayerList.java.patch:
```diff
@@ -24,9 +24,19 @@
+       p_11263_.m_8952_().m_12789_(p_11263_);
+       this.m_11273_(serverlevel1.m_6188_(), p_11263_);
+       this.f_11195_.m_129929_();
+-      MutableComponent mutablecomponent;
+-      if (p_11263_.m_36316_().getName().equalsIgnoreCase(s)) {
+-         mutablecomponent = new TranslatableComponent("multiplayer.player.joined", p_11263_.m_5446_());
+-      } else {
+-         mutablecomponent = new TranslatableComponent("multiplayer.player.joined.renamed", p_11263_.m_5446_(), s);
+-      }
+-
+-      this.m_11264_(mutablecomponent.m_130940_(ChatFormatting.YELLOW), ChatType.SYSTEM, Util.f_137441_);
++      net.minecraftforge.common.ForgeHooks.onJoinMessageFormat(p_11263_,this, s);
```
Same here, forge hook is responsible for providing the message, not sending it as well. Also, this could be made into a much smaller patch by just checking if the mutable component is not null and then doing an else if.

ChampionAsh5357 commented on src/main/java/net/minecraftforge/common/ForgeHooks.java:
```diff
@@ -1241,4 +1240,38 @@ public static void onEntityEnterSection(Entity entity, long packedOldPos, long p
+
+        if(event.isCanceled()) return;
+
+        MutableComponent mutablecomponent = event.getJoinMessage();
+        if(mutablecomponent == null) {
+            if (cachedName == null || player.getGameProfile().getName().equalsIgnoreCase(cachedName)) {
+                mutablecomponent = new TranslatableComponent("multiplayer.player.joined", player.getDisplayName());
+            } else {
+                mutablecomponent = new TranslatableComponent("multiplayer.player.joined.renamed", player.getDisplayName(), cachedName);
+            }
+            mutablecomponent = mutablecomponent.withStyle(ChatFormatting.YELLOW);
+        }
+
+        playerList.broadcastMessage(mutablecomponent, ChatType.SYSTEM, Util.NIL_UUID);
```
Already explained, you provide to the patch, not execute logic for it.

ChampionAsh5357 commented on src/main/java/net/minecraftforge/common/ForgeHooks.java:
```diff
@@ -1241,4 +1240,38 @@ public static void onEntityEnterSection(Entity entity, long packedOldPos, long p
+        MinecraftForge.EVENT_BUS.post(event);
+
+        if(event.isCanceled()) return;
+
+        MutableComponent mutableComponent = event.getLeaveMessage();
+        if(mutableComponent == null) {
+            mutableComponent = new TranslatableComponent("multiplayer.player.left", player.getDisplayName()).withStyle(ChatFormatting.YELLOW);
+        }
+
+        server.getPlayerList().broadcastMessage(mutableComponent, ChatType.SYSTEM, Util.NIL_UUID);
```
Already explained, you provide to the patch, not execute logic for it.

ChampionAsh5357 commented on src/main/java/net/minecraftforge/event/ForgeEventFactory.java:
```diff
@@ -263,6 +264,22 @@ public static Component getPlayerTabListDisplayName(Player player)
         return event.getDisplayName();
     }
 
+    @Nullable
+    public static MutableComponent getPlayerJoinMessage(Player player)
+    {
+        PlayerEvent.JoinMessageFormat event = new PlayerEvent.JoinMessageFormat(player);
+        MinecraftForge.EVENT_BUS.post(event);
+        return event.isCanceled() ? null : event.getJoinMessage();
+    }
+
+    @Nullable
+    public static MutableComponent getPlayerLeaveMessage(Player player)
+    {
+        PlayerEvent.LeaveMessageFormat event = new PlayerEvent.LeaveMessageFormat(player);
+        MinecraftForge.EVENT_BUS.post(event);
+        return  event.isCanceled() ? null : event.getLeaveMessage();
+    }
+
```
What are the point of these if they aren't used? The best way to do this is to get the original mutable component and pass it in as part of the event and then check the event before sending.

ChampionAsh5357 reviewed (changes requested):
> Also, if this pr is accepted, can I also backport the feature to 1.16.5 or is this too much of a change, for the LTS version?

Yes you can. This doesn't cause any breaking changes so it can be backported.

MojangPlsFix committed:
Use ForgeHooks like they should be

MojangPlsFix committed:
remove unused methods

MojangPlsFix committed:
remove unused import

MojangPlsFix committed:
fix java doc

MojangPlsFix commented:
I now changed the Forge Hooks to only return the Join/Leave Messages and do the default handling and removed the Methods in the ForgeEventFactory, since I forgot them to remove in my last commits.

MojangPlsFix commented:
P.s. I hope I am not being annoying requesting the review, if I shouldn't do that please just let me know about it.

ChampionAsh5357 commented on src/main/java/net/minecraftforge/common/ForgeHooks.java:
```diff
@@ -1241,4 +1240,38 @@ public static void onEntityEnterSection(Entity entity, long packedOldPos, long p
+        if(event.isCanceled()) return null;
+
+        MutableComponent mutablecomponent = event.getJoinMessage();
+        if(mutablecomponent == null) {
```
Braces on new line.

ChampionAsh5357 commented on src/main/java/net/minecraftforge/common/ForgeHooks.java:
```diff
@@ -1241,4 +1240,38 @@ public static void onEntityEnterSection(Entity entity, long packedOldPos, long p
+
+        MutableComponent mutablecomponent = event.getJoinMessage();
+        if(mutablecomponent == null) {
+            if (cachedName == null || player.getGameProfile().getName().equalsIgnoreCase(cachedName)) {
```
Doesn't need null check, `#equalsIgnoreCase` already handles that.

ChampionAsh5357 replied:
Braces on new line.

ChampionAsh5357 commented on src/main/java/net/minecraftforge/common/ForgeHooks.java:
```diff
@@ -1241,4 +1240,38 @@ public static void onEntityEnterSection(Entity entity, long packedOldPos, long p
+
+        MutableComponent mutablecomponent = event.getJoinMessage();
+        if(mutablecomponent == null) {
+            if (cachedName == null || player.getGameProfile().getName().equalsIgnoreCase(cachedName)) {
+                mutablecomponent = new TranslatableComponent("multiplayer.player.joined", player.getDisplayName());
+            } else {
+                mutablecomponent = new TranslatableComponent("multiplayer.player.joined.renamed", player.getDisplayName(), cachedName);
+            }
```
This could probably be reduced to a ternary operator, but since it's just a movement of vanilla code, this is more of an optional thing.

ChampionAsh5357 commented on src/main/java/net/minecraftforge/common/ForgeHooks.java:
```diff
@@ -1241,4 +1240,38 @@ public static void onEntityEnterSection(Entity entity, long packedOldPos, long p
+        if(mutablecomponent == null) {
+            if (cachedName == null || player.getGameProfile().getName().equalsIgnoreCase(cachedName)) {
+                mutablecomponent = new TranslatableComponent("multiplayer.player.joined", player.getDisplayName());
+            } else {
```
Else on new line, braces on new line.

ChampionAsh5357 commented on src/main/java/net/minecraftforge/common/ForgeHooks.java:
```diff
@@ -1241,4 +1240,38 @@ public static void onEntityEnterSection(Entity entity, long packedOldPos, long p
+        if(event.isCanceled()) return null;
+
+        MutableComponent mutableComponent = event.getLeaveMessage();
+        if(mutableComponent == null) {
```
Braces on new line.

ChampionAsh5357 reviewed (changes requested):
> P.s. I hope I am not being annoying requesting the review, if I shouldn't do that please just let me know about it.

I don't mind if people do so or if they ping me on discord for a review, I'll get around to it either way when I can.

MojangPlsFix committed:
fix code style

ChampionAsh5357 commented on src/main/java/net/minecraftforge/common/ForgeHooks.java:
```diff
@@ -51,7 +51,11 @@
 import it.unimi.dsi.fastutil.longs.LongSet;
+import net.minecraft.Util;
 import net.minecraft.advancements.Advancement;
+import net.minecraft.network.chat.*;
```
Avoid wildcard imports.

MojangPlsFix replied:
should be fixed now

ChampionAsh5357 reviewed (changes requested):
Just one more syntax highlight I missed; otherwise it should be good. I'll do one more round of testing to make sure.

MojangPlsFix committed:
replace .* imports thx intellij

ChampionAsh5357 commented on src/test/java/net/minecraftforge/debug/chat/JoinLeaveMessageTest.java:
```diff
@@ -0,0 +1,40 @@
+
+@Mod("client_join_leave_message_test")
+@Mod.EventBusSubscriber
+public class JoinLeaveMessageTest {
```
New line braces

ChampionAsh5357 commented on src/test/java/net/minecraftforge/debug/chat/JoinLeaveMessageTest.java:
```diff
@@ -0,0 +1,40 @@
+public class JoinLeaveMessageTest {
+
+    @SubscribeEvent
+    public static void onPlayerJoinChat(PlayerEvent.JoinMessageFormat event) {
```
New line braces

ChampionAsh5357 commented on src/test/java/net/minecraftforge/debug/chat/JoinLeaveMessageTest.java:
```diff
@@ -0,0 +1,40 @@
+    }
+
+    @SubscribeEvent
+    public static void onPlayerLeaveChat(PlayerEvent.LeaveMessageFormat event) {
```
New line braces

ChampionAsh5357 reviewed (changes requested):
Tested again and it seems to work. Just noticing more formatting errors as I go through this, sorry about that. This should be the last one. Tested and it seems to work.

MojangPlsFix committed:
comply with the coding standard

MojangPlsFix commented:
> Tested again and it seems to work. Just noticing more formatting errors as I go through this, sorry about that. This should be the last one. Tested and it seems to work.

No problem, I should have fixed it now. Regarding your coding style, is the formatter[ ForgeEssentials](https://github.com/ForgeEssentials/ForgeEssentials/blob/1.12.2/develop/misc/intellijformat.xml) Formatter still valid? Then it would be easier for you and your team the next time, if I do not make any code style errors in my next PRs.

ChampionAsh5357 reviewed (approved):
Since I've never heard of it, I have no idea. Most likely though, the formatting has not changed since then.

MojangPlsFix committed:
Merge branch '1.18.x' into 1.18.x

diesieben07 commented on patches/minecraft/net/minecraft/server/players/PlayerList.java.patch:
```diff
@@ -24,9 +24,20 @@
+       p_11263_.m_8952_().m_12789_(p_11263_);
+       this.m_11273_(serverlevel1.m_6188_(), p_11263_);
+       this.f_11195_.m_129929_();
+-      MutableComponent mutablecomponent;
+-      if (p_11263_.m_36316_().getName().equalsIgnoreCase(s)) {
+-         mutablecomponent = new TranslatableComponent("multiplayer.player.joined", p_11263_.m_5446_());
+-      } else {
+-         mutablecomponent = new TranslatableComponent("multiplayer.player.joined.renamed", p_11263_.m_5446_(), s);
+-      }
++      MutableComponent mutablecomponent = net.minecraftforge.common.ForgeHooks.getJoinMessage(p_11263_,s);
++      if(mutablecomponent != null) this.m_11264_(mutablecomponent, ChatType.SYSTEM, Util.f_137441_);
+ 
+-      this.m_11264_(mutablecomponent.m_130940_(ChatFormatting.YELLOW), ChatType.SYSTEM, Util.f_137441_);
```
This whole patch could be simpler. Instead of moving the `MutableComponent` construction out of here into ForgeHooks, pass the already computed `mutablecomponent` into `getJoinMessage` and pre-populate it in the event. Then mods can modify it at will and produce new one.
As a result this patch becomes much smaller.

diesieben07 commented on src/main/java/net/minecraftforge/common/ForgeHooks.java:
```diff
@@ -59,7 +59,16 @@
 import javax.annotation.Nullable;
 
 import it.unimi.dsi.fastutil.longs.LongSet;
+import net.minecraft.Util;
```
Please do not reorganize imports.
Please update the license header in this file

diesieben07 commented on src/main/java/net/minecraftforge/common/ForgeHooks.java:
```diff
@@ -1259,6 +1263,44 @@ public static void onEntityEnterSection(Entity entity, long packedOldPos, long p
+        PlayerEvent.JoinMessageFormat event = new PlayerEvent.JoinMessageFormat(player);
+        MinecraftForge.EVENT_BUS.post(event);
+
+        if(event.isCanceled())
```
Formatting: use braces, space after if

diesieben07 commented on src/main/java/net/minecraftforge/common/ForgeHooks.java:
```diff
@@ -1259,6 +1263,44 @@ public static void onEntityEnterSection(Entity entity, long packedOldPos, long p
+            return null;
+
+        MutableComponent mutablecomponent = event.getJoinMessage();
+        if(mutablecomponent == null)
```
space after if

diesieben07 commented on src/main/java/net/minecraftforge/common/ForgeHooks.java:
```diff
@@ -1259,6 +1263,44 @@ public static void onEntityEnterSection(Entity entity, long packedOldPos, long p
+        PlayerEvent.LeaveMessageFormat event = new PlayerEvent.LeaveMessageFormat(player);
+        MinecraftForge.EVENT_BUS.post(event);
+
+        if(event.isCanceled())
```
space after if, use braces

diesieben07 commented on src/main/java/net/minecraftforge/common/ForgeHooks.java:
```diff
@@ -1259,6 +1263,44 @@ public static void onEntityEnterSection(Entity entity, long packedOldPos, long p
+            return null;
+
+        MutableComponent mutableComponent = event.getLeaveMessage();
+        if(mutableComponent == null)
```
space after if, please use a more meaningful variable name than "mutableComponent"

diesieben07 commented on src/main/java/net/minecraftforge/event/entity/player/PlayerEvent.java:
```diff
@@ -218,6 +220,80 @@ public void setDisplayName(@Nullable Component displayName)
+     * JoinMessageFormat is fired when a player's join message is retrieved.<br>
+     * This event is fired whenever a player's join message is retrieved, when a player tries to join
+     * <br>
+     * This event is fired via the {@link net.minecraftforge.common.ForgeHooks#getJoinMessage(Player, String)}.<br>
```
This information is entirely irrelevant.

diesieben07 commented on src/main/java/net/minecraftforge/event/entity/player/PlayerEvent.java:
```diff
@@ -218,6 +220,80 @@ public void setDisplayName(@Nullable Component displayName)
+     * LeaveMessageFormat is fired when a player's leave message is retrieved.<br>
+     * This event is fired whenever a player's leave message is retrieved, when a player tries to leave
+     * <br>
+     * This event is fired via the {@link net.minecraftforge.common.ForgeHooks#getLeaveMessage(Player, MinecraftServer)}.<br>
```
This information is entirely irrelevant

diesieben07 commented on src/test/java/net/minecraftforge/debug/chat/JoinLeaveMessageTest.java:
```diff
@@ -0,0 +1,42 @@
+/*
+ * Minecraft Forge
+ * Copyright (c) 2016-2021.
```
Please update the license header

diesieben07 reviewed (changes requested):
I know I am late and this is already assigned, but I noticed some issues here.
In addition to my comment on the patch, please ensure the formatting is correct.

LexManos commented:
All of these need to be simplified by passing thee original message from vanilla into the event. This way we never run into the issues where we desync from what vanilla changes.

Not the mention the plethora of formatting issues.
